conversion/resource_generation.go

```diff
@@ -0,0 +1,401 @@
+}
+
+type ValidateOrCreateResourcesInterface interface {
+	ValidateOrCreateResourcesForShardedMigration(ctx context.Context, projectId string, instanceName string, validateOnly bool, region string, sourceProfile profiles.SourceProfile) error
+}
+
+type ValidateOrCreateResourcesImpl struct {
+	ResourceGenerator ResourceGenerationInterface
+	RunParallel       common.RunParallelTasksInterface[*ConnectionProfileReq, *ConnectionProfileReq]
+}
+
+type ValidateResourcesInterface interface {
+	ValidateResourceGeneration(ctx context.Context, projectId string, instanceId string, sourceProfile profiles.SourceProfile, conv *internal.Conv) error
+}
+
+type ValidateResourcesImpl struct {
+	SpAcc                     spanneraccessor.SpannerAccessor
+	SpInstanceAdmin           spinstanceadmin.InstanceAdminClient
+	ValidateOrCreateResources ValidateOrCreateResourcesInterface
+}
+
+func NewValidateResourcesImpl(spAcc spanneraccessor.SpannerAccessor, spInstanceAdmin spinstanceadmin.InstanceAdminClient, dsAcc datastream_accessor.DatastreamAccessor, dsClient ds.DatastreamClient, storageAcc storageaccessor.StorageAccessor, storageClient storageclient.StorageClient) *ValidateResourcesImpl {
+	return &ValidateResourcesImpl{
+		SpAcc:                     spAcc,
+		SpInstanceAdmin:           spInstanceAdmin,
+		ValidateOrCreateResources: NewValidateOrCreateResourcesImpl(dsAcc, dsClient, storageAcc, storageClient),
+	}
+}
+
+func NewValidateOrCreateResourcesImpl(dsAcc datastream_accessor.DatastreamAccessor, dsClient ds.DatastreamClient, storageAcc storageaccessor.StorageAccessor, storageClient storageclient.StorageClient) *ValidateOrCreateResourcesImpl {
+	return &ValidateOrCreateResourcesImpl{
+		ResourceGenerator: NewResourceGenerationImpl(dsAcc, dsClient, storageAcc, storageClient),
+		RunParallel:       &common.RunParallelTasksImpl[*ConnectionProfileReq, *ConnectionProfileReq]{},
+	}
+}
+
+func NewResourceGenerationImpl(dsAcc datastream_accessor.DatastreamAccessor, dsClient ds.DatastreamClient, storageAcc storageaccessor.StorageAccessor, storageClient storageclient.StorageClient) *ResourceGenerationImpl {
+	return &ResourceGenerationImpl{
+		DsAcc:         dsAcc,
+		DsClient:      dsClient,
+		StorageAcc:    storageAcc,
+		StorageClient: storageClient,
+	}
+}
+
+// Method to validate if in a minimal downtime migration, required resources can be generated
+func (v *ValidateResourcesImpl) ValidateResourceGeneration(ctx context.Context, projectId string, instanceId string, sourceProfile profiles.SourceProfile, conv *internal.Conv) error {
+	spannerRegion, err := v.SpAcc.GetSpannerLeaderLocation(ctx, v.SpInstanceAdmin, "projects/"+projectId+"/instances/"+instanceId)
+	if err != nil {
+		err = fmt.Errorf("unable to fetch Spanner Region: %v", err)
+		return err
+	}
+	conv.SpRegion = spannerRegion
+	err = v.ValidateOrCreateResources.ValidateOrCreateResourcesForShardedMigration(ctx, projectId, instanceId, true, spannerRegion, sourceProfile)
```

Comment:
This looks odd - `ValidateResource` method has a `ValidateOrCreate` inside it. Is this a naming issue or does the code require restructuring?

Reply:
Answering both the last and this comment here:
I have separated out things that need to only happen for validate in "ValidateResourceGeneration", but in "ValidateOrCreateResourcesForShardedMigration" the only difference is of a boolean value we send to the backend which is why if I bread them, it would be the same code with one validateOnly set as true and the other set as false

Reply:
I dug into this a little bit deeper, and I think the root of the problem is that the Datastream API itself is very awkward. The `CreateConnectionProfileRequest` contains the `validateOnly` flag which makes it function effectively like a `ValidateConnectionProfileRequest` object. 

This same awkwardness is percolating in our codebase, which is why we have these kind of `ValidateOrCreate` methods, since we are effectively setting a parameter in the API request.

Had they been two APIs, we would have not done so.

I think its okay for us to move ahead with existing structure given the nature of the API, but it sure would have been cleaner if we were able to separate out these flows!

Reply:
Okay thankyou